### PR TITLE
Adds `Progressing` status to `BackupBucketsReady` seed condition 

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
Adds a Progressing status to the `BackupBucketsReady` seed condition. When a BackupBucket that refers to a seed becomes erroneous, the `BackupBucketsReady` condition will first be set to `Progressing`, if the configured condition threshold has not expired. Once the threshold expires, the condition will be set to `False`

**Which issue(s) this PR fixes**:
Part of #4722

**Special notes for your reviewer**:

**Release note**:

```other operator
The `BackupBucketsReady` condition for `Seed`s will first be set to `Progressing` instead of being directly set to `False` when a `BackupBucketsReady` condition threshold is specified in the `controllers.seedBackupBucketsCheck.conditionThresholds` configuration for the gardener controller manager and that threshold has not expired yet. 
```

### CODE REVIEW

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go [hunk 1]
```diff
@@ -106,6 +112,13 @@ func (b *backupBucketReconciler) Reconcile(ctx context.Context, req reconcile.Re
 		erroneousBackupBuckets []backupBucketInfo
 	)
 	for _, bb := range backupBucketList.Items {
+		// not needed for real client, but fake client doesn't support field selector
+		// see https://github.com/kubernetes-sigs/controller-runtime/issues/1376
+		// could be solved by switching from fake client to real client against envtest
+		if bb.Spec.SeedName == nil || *bb.Spec.SeedName != seed.Name {
+			continue
+		}
+
```
Comment (shafeeqes) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go]: Maybe you can do something like this: https://github.com/gardener/gardener/blob/5b3456739b23d2009b69e65b642892c507842a6a/pkg/controllermanager/controller/bastion/add_test.go#L213-L216

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control_test.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+							Reason:  "BackupBucketsError",
+							Status:  gardencorev1beta1.ConditionFalse,
+							Type:    gardencorev1beta1.SeedBackupBucketsReady,
+						},
+					}
+					Expect(c.Update(ctx, seed)).To(Succeed())
+
+					matchExpectedCondition = MatchFields(IgnoreExtras, Fields{
+						"Message": Equal("The following BackupBuckets have issues:\n* Name: 1, Error: foo error"),
+						"Type":    Equal(gardencorev1beta1.SeedBackupBucketsReady),
```
Comment (shafeeqes) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control_test.go:272]: Nit: 
```suggestion
						"Type":    Equal(gardencorev1beta1.SeedBackupBucketsReady),
						"Status":  Equal(gardencorev1beta1.ConditionFalse),
```

**Review by shafeeqes — COMMENTED** [review]
Thank you!
I have only few minor comments.

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go (hunk 1 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go]: Thanks, that seems like a nice approach. I can probably also use it for the extensions check controller, but in a separate PR.

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go [hunk 3]
```diff
@@ -57,41 +58,46 @@ func (c *Controller) backupBucketUpdate(oldObj, newObj interface{}) {
...
+	clock        clock.Clock
 }
 
 type backupBucketInfo struct {
 	name     string
 	errorMsg string
 }
 
-func (b *backupBucketInfo) String() string {
+func (b *backupBucketInfo) string() string {
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go]: Q: what was the motivation for this change?

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go [hunk 4]
```diff
@@ -18,26 +18,27 @@ import (
...
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	"github.com/gardener/gardener/pkg/apis/core"
 	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
 	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
 )
 
-const seedBackupBucketReconcilerName = "backupbucket"
+const seedBackupBucketsCheckReconcilerName = "backupbuckets-check"
```
Comment (ialidzhikov) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go:33]: Nit: The reconciler name can be updated in https://github.com/gardener/gardener/blob/195189a61500b1010d9555c816755a91c0747e7c/docs/concepts/controller-manager.md#L189 as well.

**Review by ialidzhikov — COMMENTED** [review]
lgtm in general

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go (hunk 3 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go]: I guess the idea initially was for the backupBucketInfo to implement the `Stringer` interface, but it wasn't used as such anywhere, and instead the `.String()` function was called directly here: https://github.com/gardener/gardener/blob/b49fb4e45c6de81bfcefbe0cbd0612577518d9e9/pkg/controllermanager/controller/seed/seed_backup_bucket_reconcile.go#L122

Anyway, I will fix it so that the implementation is done correctly and `Stringer` is used.

File: test/integration/controllermanager/seed/backupbucketscheck/backupbucketscheck_test.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+package backupbucketscheck_test
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/gstruct"
+	gomegatypes "github.com/onsi/gomega/types"
+	v1 "k8s.io/api/core/v1"
```
Comment (rfranzke) [inline test/integration/controllermanager/seed/backupbucketscheck/backupbucketscheck_test.go]: Use `corev1`

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go [hunk 6]
```diff
@@ -57,41 +58,46 @@ func (c *Controller) backupBucketUpdate(oldObj, newObj interface{}) {
 		return
 	}
 
-	if !apiequality.Semantic.DeepEqual(oldBackupBucket.Status, newBackupBucket.Status) || !apiequality.Semantic.DeepEqual(oldBackupBucket.Spec, newBackupBucket.Spec) {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go:60]: Why did you change this condition?

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go [hunk 7]
```diff
@@ -57,41 +58,46 @@ func (c *Controller) backupBucketUpdate(oldObj, newObj interface{}) {
...
-func (b *backupBucketReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+func (b *backupBucketsCheckReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
 	log := logf.FromContext(ctx)
 
 	seed := &gardencorev1beta1.Seed{}
 	if err := b.gardenClient.Get(ctx, req.NamespacedName, seed); err != nil {
 		if apierrors.IsNotFound(err) {
 			log.V(1).Info("Object is gone, stop reconciling")
-			return reconcile.Result{}, nil
+			return reconcileResult(nil)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go]: Can you drop this `reconcileResult` function and revert related changes? We don't have this anywhere else and it's unnecessarily "simplifying" the code

File: pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go (hunk 6 above)
Comment (plkokanov) [reply pkg/controllermanager/controller/seed/seed_backup_bucket_check_control.go:60]: The reconciler only checks the `backupBucket.status.lastError` and `backupBucket.status.lastError.description` so I felt like there was no reason to enqueue it in case of other changes. Additionally, now the the reconciliation will be requeued based on the `syncPeriod` setting and I felt like there was no reason to additionally enqueue it only based on `backupEntry.spec` changes.

**Comment by shafeeqes** [comment]
> @plkokanov: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-gardener-apidiff	[7820a24](https://github.com/gardener/gardener/commit/7820a245f83778cbd391398a000e142d5b1ada47)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6587/pull-gardener-apidiff/1565751885325930496)	false	`/test pull-gardener-apidiff`
> pull-gardener-e2e-kind	[7820a24](https://github.com/gardener/gardener/commit/7820a245f83778cbd391398a000e142d5b1ada47)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6587/pull-gardener-e2e-kind/1565751885468536832)	true	`/test pull-gardener-e2e-kind`
> [Full PR test history](https://prow.gardener.cloud/pr-history?org=gardener&repo=gardener&pr=6587). [Your PR dashboard](https://prow.gardener.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Aplkokanov). [Command help for this repository](https://prow.gardener.cloud/command-help?repo=gardener%2Fgardener). Please help us cut down on flakes by linking this test failure to an [open flake report](https://github.com/gardener/gardener/issues?q=is:issue+is:open+label:kind/flake) or filing a [new flake report](https://github.com/gardener/gardener/issues/new?template=flaking-test.md&title=%5BFlaky+Test%5D+FLAKING+TEST%2FSUITE) if you can't find an existing one. Also see our [testing guideline](https://github.com/gardener/gardener/blob/master/docs/development/testing.md) for how to avoid and hunt flakes.

Opened https://github.com/gardener/gardener/issues/6623
/retest
